### ISSUE TITLE
Add release atteststions

### ISSUE THREAD

**Issue by orf** [astral-sh/python-build-standalone#343]
Hello!

First, thanks for this fantastic project - it’s a great help to the community and ecosystem at large.

I would like to suggest adding [artifact attestations ](https://docs.github.com/en/actions/security-for-github-actions/using-artifact-attestations/using-artifact-attestations-to-establish-provenance-for-builds) to the releases.

This would enable supply chain verification for these builds, and provide a layer of validation above just verifying the signature.

it’s pretty simple to add: just a single step, with no configuration or changes required in the binary itself.

what do you think?

**Comment by zanieb** [comment]
Yeah these seem reasonable, were you interested in contributing this?

**Comment by orf** [comment]
Absolutely! I’m currently travelling and away from my laptop so it will have to be next week, but I can take a look at it when I’m back.

**Comment by chludwig-haufe** [comment]
I see there is an open PR #371, but no further activity. Is there anything blocking the creation of the attestations?

Since I didn't see this mentioned in the documentation: How is the integrity of the Python sources and dependencies used in the build verified?

**Comment by indygreg** [comment]
We pin and verify the sha256 and file size of all downloaded assets. See downloads.py.

**Comment by zanieb** [comment]
@chludwig-haufe there's a blocking comment at https://github.com/indygreg/python-build-standalone/pull/371#issuecomment-2416938621

**Comment by orf** [comment]
> @chludwig-haufe there's a blocking comment at [#371 (comment)](https://github.com/indygreg/python-build-standalone/pull/371#issuecomment-2416938621)

Yep! I have this PR in my backlog, but some personal stuff has come up.

We should indeed only attest on main, adding the relevant `if: ` blocks should be all that is needed.

**Comment by zanieb** [comment]
I presume we need to attest the artifacts in `.github/workflows/release.yml` too? We create several derived artifacts.

**Comment by chludwig-haufe** [comment]
Please pardon my ignorance (I am not an active user of GitHub actions) – but [Verifying artifact attestations with the GitHub CLI](https://docs.github.com/en/actions/security-for-github-actions/using-artifact-attestations/using-artifact-attestations-to-establish-provenance-for-builds#verifying-artifact-attestations-with-the-github-cli) made me assume I could verify the integrity of a downloaded package as follows:

~~~shell
$ curl -LO https://github.com/astral-sh/python-build-standalone/releases/download/20250115/cpython-3.10.16+20250115-aarch64-apple-darwin-install_only.tar.gz
$ export GH_TOKEN="GitHub personal access token w/ read privileges"
$ gh attestation verify cpython-3.10.16+20250115-aarch64-apple-darwin-install_only.tar.gz -R astral-sh/python-build-standalone
Loaded digest sha256:21dba90e7a0b879fdc475411fc8c39e0567c6f5c92112956b63de7f83ee433ad for file://cpython-3.10.16+20250115-aarch64-apple-darwin-install_only.tar.gz
✗ Loading attestations from GitHub API failed

Error: failed to fetch attestations from astral-sh/python-build-standalone: HTTP 404: Not Found (https://api.github.com/repos/astral-sh/python-build-standalone/attestations/sha256:21dba90e7a0b879fdc475411fc8c39e0567c6f5c92112956b63de7f83ee433ad?per_page=30)
~~~

What am I missing?

**Comment by zanieb** [comment]
The `-install_only` archives are derived, as I mentioned in https://github.com/astral-sh/python-build-standalone/issues/343#issuecomment-2593554836

**Comment by zanieb** [comment]
Let's just re-open this until we reach a point where these are validated as functional.

**Comment by chludwig-haufe** [comment]
> The `-install_only` archives are derived, as I mentioned in [#343 (comment)](https://github.com/astral-sh/python-build-standalone/issues/343#issuecomment-2593554836)

Ah, I didn't make the connection that your comment referred to installable packages, too. I had assumed you meant SBOMs and similar metadata. (I am obviously not familiar with your build pipeline.) Thanks for the clarification.

**Comment by chludwig-haufe** [comment]
I can verify, e..g., `cpython-3.10.16+20250115-aarch64-apple-darwin-pgo+lto-full.tar.zst`.
On the other hand, the verification of `cpython-3.10.16+20250115-aarch64-unknown-linux-gnu-lto-full.tar.zst` fails with a 404.

Looking at [.github/workflows/linux.yml](https://github.com/astral-sh/python-build-standalone/blob/main/.github/workflows/linux.yml), I suspect the former artifact is generated by `build-1` and the latter by `build-0`:  `build-1` includes a ["Generate Attestation" step in line 326](https://github.com/astral-sh/python-build-standalone/blob/fc29abe434cdbf48426bc1c1270232b898f2f5f2/.github/workflows/linux.yml#L326), whereas `build-0` has no such step.

**Comment by samypr100** [comment]
@chludwig-haufe can you try again on the latest release?

**Comment by chludwig-haufe** [comment]
I downloaded and successfully verified all(?) 564 artifacts matching the shell pattern `cpython*.tar.{gz,zst}` - looks good!

**Comment by zanieb** [comment]
Thank you!
